From 94fe4a0a1efcb979673d33423c7f6edbd6e6b131 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sun, 10 Nov 2019 02:46:14 +0100
Subject: [PATCH 1/1] efi_loader: unaligned copies in device path functions

*((struct efi_device_path *)buf) = END;

is compiled as a call to  memcpy() in most cases. Due to packing the
address of buf will be unaligned in device path functions. To be sure that
memcpy() is always used replace the lines by an explicit invocation of
memcpy().

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_device_path.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/lib/efi_loader/efi_device_path.c b/lib/efi_loader/efi_device_path.c
index afef4e8450..0418106f50 100644
--- a/lib/efi_loader/efi_device_path.c
+++ b/lib/efi_loader/efi_device_path.c
@@ -784,7 +784,7 @@ struct efi_device_path *efi_dp_from_part(struct blk_desc *desc, int part)
 
 	buf = dp_part_fill(buf, desc, part);
 
-	*((struct efi_device_path *)buf) = END;
+	memcpy(buf, &END, sizeof(END));
 
 	return start;
 }
@@ -870,7 +870,7 @@ struct efi_device_path *efi_dp_from_file(struct blk_desc *desc, int part,
 	path_to_uefi(fp->str, path);
 	buf += fpsize;
 
-	*((struct efi_device_path *)buf) = END;
+	memcpy(buf, &END, sizeof(END));
 
 	return start;
 }
@@ -912,7 +912,7 @@ struct efi_device_path *efi_dp_from_eth(void)
 	buf = &ndp[1];
 #endif
 
-	*((struct efi_device_path *)buf) = END;
+	memcpy(buf, &END, sizeof(END));
 
 	return start;
 }
@@ -939,7 +939,7 @@ struct efi_device_path *efi_dp_from_mem(uint32_t memory_type,
 	mdp->end_address = end_address;
 	buf = &mdp[1];
 
-	*((struct efi_device_path *)buf) = END;
+	memcpy(buf, &END, sizeof(END));
 
 	return start;
 }
-- 
2.24.0

